packages/kbn-config-schema/src/internals/index.ts

```diff
@@ -226,198 +171,197 @@ export const internals = Joi.extend([
       // > strings that can be converted to numbers)
       const coercedValue: any = typeof value === 'string' ? Number(value) : value;
       if (typeof coercedValue !== 'number' || isNaN(coercedValue)) {
-        return this.createError('number.base', { value }, state, options);
+        return {
+          errors: [error('number.base')],
+        };
       }
 
-      return value;
+      return { value };
     },
-    rules: [anyCustomRule],
   },
   {
-    name: 'object',
-
+    type: 'object',
     base: Joi.object(),
-    coerce(value: any, state: State, options: ValidationOptions) {
+    coerce(value: any, { error, prefs }) {
       if (value === undefined || isPlainObject(value)) {
-        return value;
+        return { value };
       }
 
-      if (options.convert && typeof value === 'string') {
+      if (prefs.convert && typeof value === 'string') {
         try {
           const parsed = JSON.parse(value);
-          if (isPlainObject(parsed)) {
-            return parsed;
-          }
-          return this.createError('object.base', { value: parsed }, state, options);
+          return { value: parsed };
         } catch (e) {
-          return this.createError('object.parse', { value }, state, options);
+          return { errors: [error('object.parse')] };
         }
       }
 
-      return this.createError('object.base', { value }, state, options);
+      return { errors: [error('object.base')] };
     },
-    rules: [anyCustomRule],
   },
   {
-    name: 'map',
-
-    coerce(value: any, state: State, options: ValidationOptions) {
+    type: 'array',
+    base: Joi.array(),
+    coerce(value: any, { error, prefs }) {
+      if (value === undefined || Array.isArray(value)) {
+        return { value };
+      }
+      if (prefs.convert && typeof value === 'string') {
+        try {
+          // ensuring that the parsed object is an array is done by the base's validation
+          return { value: JSON.parse(value) };
+        } catch (e) {
+          return {
+            errors: [error('array.parse')],
+          };
+        }
+      }
+      return {
+        errors: [error('array.base')],
+      };
+    },
+  },
+  {
+    type: 'map',
+    coerce(value, { error, prefs }) {
       if (value === undefined) {
-        return value;
+        return { value };
       }
       if (isPlainObject(value)) {
-        return new Map(Object.entries(value));
+        return { value: new Map(Object.entries(value)) };
       }
-      if (options.convert && typeof value === 'string') {
+      if (prefs.convert && typeof value === 'string') {
```

Comment:
Are we rely on `joi` defaults for `convert`? Maybe we shouldn't?

Reply:
`prefs.convert` defaults to `true`, and `config-schema` doesn't provide any way to change this value, so effectively this part of the condition could be removed. I thought it was better to keep it, just to kinda preserve consistency with how the underlying library is working. But maybe not.